ARROW-5893: [C++][Python][GLib][Ruby][MATLAB][R] Remove arrow::Column class

wesm opened the pull request:
This completely removes `arrow::Column` from the C++ library and adapts the Python bindings to follow, to help assist with mailing list discussion.

There are other places that this would touch:

- [x] GLib
- [x] Ruby
- [x] MATLAB
- [x] R

You can see the evident reduction of boilerplate and simplification in many places -- it is particularly pronounced in parquet/arrow/arrow-reader-writer-test.cc. This refactor also exposed a bug where a Column's data type did not match its Field type.

jorisvandenbossche commented on cpp/src/arrow/ipc/feather.cc:
```diff
@@ -434,35 +434,33 @@ class TableReader::TableReaderImpl {
       RETURN_NOT_OK(GetColumn(i, &column));
       columns.push_back(column);
-      fields.push_back(column->field());
+      fields.push_back(::arrow::field(GetColumnName(i), column->type()));
```
(I see that the feather.fbs has no notion of nullabilty, so that answers my comment)

bkietz replied:
@jorisvandenbossche @wesm definitely out of scope for this PR, but: should user_metadata get into the read table's fields?

```suggestion
      fields.push_back(::arrow::field(GetColumnName(i), column->type(), true, {{"user_metadata"}, {user_metadata}}));
```

wesm replied:
Maybe so. Can you open a JIRA about this? The Feather format is effectively deprecated so not sure how much more energy we want to invest in it (versus having Feather be a wrapper for the primary IPC protocol)

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -307,287 +345,10 @@ def chunked_array(arrays, type=None):
-        """
-        return self.data.unique()
-
-    def flatten(self, MemoryPool memory_pool=None):
```
Do we want to keep this on the Array/ChunkedArray as well? (similarly to how you moved Column.cast to ChunkedArray.cast)

wesm replied:
Yes, will do

kevingurney commented:
@wesm - This looks like a nice API change.

Since the MATLAB `featherread`/`featherwrite` interfaces currently use `arrow::Column` under the hood (as you pointed out), should we try to include the MATLAB code changes as part of this pull request, or as a separate pull request?

Either way, @rdmello and I are happy to help with making those changes.

wesm commented:
@kevingurney feel free to post a link to a branch with any changes and I can cherry-pick them into this PR

kevingurney commented:
Sounds good - thanks!

wesm commented:
@kou @shiro615 I have just performed the changes to the GLib C++ code, but it would be much quicker for one of you to fix the unit tests. I wasn't sure how to get the `GARROW_AVAILABLE_IN_...` macros to work

The Ruby library will have to have some minor modifications but it should not be too complicated

I apologies for the disruption -- if you could say whether you support this change either here or in the mailing list thread that would be helpful.

kevingurney commented:
@wesm - Changes for removing the use of `arrow::Column` under the hood in the MATLAB `featherread`/`featherwrite` interfaces can be found in the branch linked below:

- https://github.com/mathworks/arrow/tree/arrow-5893

I also updated `handle_status.cc` because of a build failure I encountered caused by recent changes to `arrow::StatusCode`.

@rdmello - Can you take a quick look through these changes?

kou commented:
@wesm No problem. I take over GLib related part.

rdmello commented:
@kevingurney @wesm thanks for doing this. I have looked through the MATLAB changes and didn't see any issues.

wesm commented:
OK, I'll cherry-pick the changes here then

kou commented:
I finished GLib part.
I'll work on Ruby part later.

I'll comment whether this change is reasonable in GLib and Ruby parts when I finish Ruby part work.

wesm commented:
Thanks @kou! And sorry for the disruption. I'll try to fix up the R package and get a passing build early in the week

kou commented:
I've finished Ruby part. And I've also fixed remained problems in C++ part.

I realized that column object is useful to build user friendly API in Ruby. So I implement column class in Ruby. Column object that has metadata (field) and data (chunked array) is useful to represent column information. We can represent column as an array table object and column index but creating a column object is natural in Ruby. Because Ruby is pure object oriented language and uses dynamic typing. 

But column object is needless in C++ and GLib. So I vote +1 to this change.

@shiro615 What do you think about this? Here are changes what I did:

GLib:

  * Changed `*_get_column()` to `*_get_column_data()`. Because "data" isn't enough information for column. Column needs metadata such as column name.
    * Removing `*_get_column()` is backward incompatible change. But we can do this for 1.0.0.
  * Removed `*_get_columns()`. Because users will be confused when `*_get_columns()` return column data instead of column.
  * Added some missing bindings such as `garrow_schema_get_field_index()` and `garrow_chunked_array_get_n_rows()` to implement column object in Ruby.

Ruby:

  * Implemented `Arrow::Column` in Ruby. It has similar API before but some APIs such as `#initialize` are backward incompatible.
  * Refactored some features to adjust `Arrow::Column` related changes.

wesm commented:
Thanks @kou! I agree that from binding to binding the Column concept may or may not make sense.

Once we build an initial version of the Data Frame API I think that we will want users to use that instead. That will take some months still

shiro615 commented:
@kou This changes looks good to me. Thanks.

wesm commented on cpp/src/arrow/table.h:
```diff
@@ -265,7 +169,10 @@ class ARROW_EXPORT Table {
+  virtual std::shared_ptr<ChunkedArray> column(int i) const = 0;
+
+  /// Return a column's field by index
+  virtual std::shared_ptr<Field> field(int i) const = 0;
```
agreed

wesm commented on cpp/src/arrow/table.h:
```diff
@@ -293,11 +200,13 @@ class ARROW_EXPORT Table {
 
   /// \brief Add column to the table, producing a new Table
-  virtual Status AddColumn(int i, const std::shared_ptr<Column>& column,
+  virtual Status AddColumn(int i, std::shared_ptr<Field> field_arg,
```
You can add a column at any position in the table

wesm commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -697,18 +690,13 @@ std::vector<int> FileReader::Impl::GetDictionaryIndices(const std::vector<int>&
-    auto dict_field = std::make_shared<::arrow::Field>(name, dict_array->type());
-    fields[idx] = dict_field;
-    columns[idx] = std::make_shared<Column>(dict_field, dict_array);
+    fields[idx] = ::arrow::field(old_schema.field(idx)->name(), columns[idx]->type());
```
done

wesm commented on python/pyarrow/feather.pxi:
```diff
@@ -98,12 +98,12 @@ cdef class FeatherReader:
             raise IndexError(i)
 
-        cdef shared_ptr[CColumn] sp_column
+        cdef shared_ptr[CArray] sp_array
```
done

wesm commented on python/pyarrow/table.pxi:
```diff
@@ -1184,17 +963,16 @@ cdef class Table(_PandasConvertible):
+            Names for the table columns. If not passed, schema must be passed
         schema : Schema, default None
-            If not passed, will be inferred from the arrays
+            Must pass this or names
```
done

wesm commented:
OK. I fixed up the R build -- my first time ever building the R package!

To future-proof the Feather API, I changed `FeatherReader::GetColumn` to return `ChunkedArray` instead of `Array`. I wasn't able to test the MATLAB change so please send a follow up PR if I broke it (sorry)

wesm commented:
+1. I'll wait for CI to run. Thanks everyone for your help on this, good team work =) I will use this PR as a change to test ARROW-5716 (attributing Co-authors in the GitHub UI when merging PRs)

nealrichardson commented on r/R/feather.R:
```diff
@@ -34,7 +34,7 @@
     num_columns = function() ipc___feather___TableReader__num_columns(self),
     GetColumnName = function(i) ipc___feather___TableReader__GetColumnName(self, i),
-    GetColumn = function(i) shared_ptr(`arrow::Column`, ipc___feather___TableReader__GetColumn(self, i)),
+    GetColumn = function(i) shared_ptr(`arrow::Array`, ipc___feather___TableReader__GetColumn(self, i)),
```
Just an observation that every other change mapped Column to ChunkedArray but this one is just Array. Should this one also be ChunkedArray?

kevingurney commented:
@wesm - I'll take a look at the `ChunkedArray` changes for the MATLAB code. I can include any changes in a separate PR like you suggested to avoid holding up merging in this PR.

wesm commented:
Hmm. Python 2.7 is failing. I'll see if I can reproduce locally in a little while

kevingurney commented:
@wesm  - I just took a closer look at your changes and they look good to me. I ran all the MATLAB `featherread`/`featherwrite` tests and they pass. Whenever Feather files with multiple chunks become a reality, then we will have to make some more changes to properly support them. Thanks!

wesm commented:
Here's builds in progress on my fork

Travis CI: https://travis-ci.org/wesm/arrow/builds/559726306
Appveyor: https://ci.appveyor.com/project/wesm/arrow/builds/26030853

wesm commented:
Forgot about the docs

wesm commented:
Hm a build ran on my Travis CI but it only compares one commit with the last. 

I just pushed a squashed build here to make sure all the builds run:

Travis CI: https://travis-ci.org/wesm/arrow/builds/559761819
Appveyor: https://ci.appveyor.com/project/wesm/arrow/builds/26032806

@nealrichardson it looks like the R Appveyor build is not taking into account changes in the C++ library, see that `arrow::Column` is in the logs here. I think it's building against master?

https://ci.appveyor.com/project/wesm/arrow/builds/26030853/job/7vn8q3l8e24t83jh?fullLog=true

wesm commented:
OK, the builds look good. I'm going to try to merge this with ARROW-5716
